File: LeetCode_ProblemSet/Problem1976.cpp
#include <climits>
#include <functional>
#include <queue>
#include <set>
#include <vector>

class Solution {
   private:
    long long INF = 1000000000000;
    int MOD = 1e9 + 7;

   public:
    int countPaths(int n, std::vector<std::vector<int>>& roads) {
        const std::vector<std::vector<long long>> adj = getAdj(n, roads);
        const std::vector<long long> dis = disjkstra(0, adj);
        std::pair<std::vector<std::vector<int>>, std::vector<int>> dag = getShortestPathDAG(roads, dis);
        return numberOfPaths(0, n - 1, dag.first, dag.second);
    }

   private:
    const std::vector<std::vector<long long>> getAdj(const int n, const std::vector<std::vector<int>>& roads) {
        std::vector<std::vector<long long>> adj(n, std::vector<long long>(n, INF));

        for (const auto& road : roads) {
            int u = road[0], v = road[1], wt = road[2];
            adj[u][v] = adj[v][u] = wt;
        }

        return adj;
    }

    const std::vector<long long> disjkstra(const int src, const std::vector<std::vector<long long>>& adj) {
        std::vector<long long> dis(adj.size(), INF);
        std::set<std::pair<long long, int>> st;

        st.insert({0, src});
        dis[src] = 0;

        while (!st.empty()) {
            auto p = *(st.begin());
            st.erase(p);
            int node = p.second;

            for (int adjNode = 0; adjNode < adj.size(); adjNode++) {
                long long wt = adj[node][adjNode];

                if (dis[adjNode] > dis[node] + wt) {
                    if (dis[adjNode] != INF) st.erase({dis[adjNode], adjNode});

                    dis[adjNode] = dis[node] + wt;
                    st.insert({dis[adjNode], adjNode});
                }
            }
        }

        return dis;
    }

    std::pair<std::vector<std::vector<int>>, std::vector<int>> getShortestPathDAG(const std::vector<std::vector<int>>& roads, const std::vector<long long>& dis) {
        std::vector<std::vector<int>> adj(dis.size());
        std::vector<int> parentCount(dis.size(), 0);

        for (const auto& road : roads) {
            int u = road[0], v = road[1], wt = road[2];

            if (dis[u] + wt == dis[v]) {
                parentCount[v]++;
                adj[u].push_back(v);

            } else if (dis[v] + wt == dis[u]) {
                parentCount[u]++;
                adj[v].push_back(u);
            }
        }

        return {adj, parentCount};
    }

    const std::vector<int> topologicalSorting(std::vector<int>& parentCount, const std::vector<std::vector<int>>& adj) {
        std::vector<int> sorted;
        std::queue<int> q;

        sorted.reserve(parentCount.size());

        for (int i = 0; i < parentCount.size(); i++) {
            if (!parentCount[i]) q.push(i);
        }

        while (!q.empty()) {
            int node = q.front();
            q.pop();

            sorted.push_back(node);

            for (const int adjNode : adj[node]) {
                parentCount[adjNode]--;

                if (!parentCount[adjNode]) q.push(adjNode);
            }
        }

        return sorted;
    }

    int numberOfPaths(int src, int dst, const std::vector<std::vector<int>>& adj, std::vector<int>& parentCount) {
        std::vector<int> sorted = topologicalSorting(parentCount, adj);
        std::vector<int> dp(parentCount.size(), 0);

        dp[dst] = 1;

        for (int i = sorted.size() - 1; i >= 0; i--) {
            for (int j = 0; j < adj[sorted[i]].size(); j++) {
                dp[sorted[i]] = (dp[sorted[i]] + dp[adj[sorted[i]][j]]) % MOD;
            }
        }

        return dp[src];
    }
};

File: LeetCode_ProblemSet/Rust/src/Problem1976.rs
use std::{collections::{BinaryHeap, VecDeque}, vec};

const INF: i128 = 1e12 as i128;
const MOD: i32 = 1e9 as i32 + 7;

impl Solution {
    pub fn count_paths(n: i32, roads: Vec<Vec<i32>>) -> i32 {
        let adj = Self::get_adj(n, &roads);
        let dis = Self::dijkstra(0, &adj);
        let (adj, mut parent_count) = Self::get_shortest_path_dag(&roads, &dis);
        let sorted = Self::topological_sorting(&mut parent_count, &adj);
        return Self::number_of_paths(0, n - 1, &adj, &sorted);
    }

    fn get_adj(n: i32, roads: &Vec<Vec<i32>>) -> Vec<Vec<i128>> {
        let mut adj = vec![vec![INF; n as usize]; n as usize];

        for road in roads {
            let u = road[0];
            let v = road[1];
            let wt = road[2];

            adj[u as usize][v as usize] = wt as i128;
            adj[v as usize][u as usize] = wt as i128;
        }

        return adj;
    }

    fn dijkstra(src: i32, adj: &Vec<Vec<i128>>) -> Vec<i128> {
        let mut dis = vec![INF; adj.len()];
        let mut st: BinaryHeap<(i128, i32)> = BinaryHeap::new();

        st.push((0, src));
        dis[src as usize] = 0;

        while !st.is_empty() {
            let p = st.pop();
            let node;
            match p {
                None => break,
                Some(p) => node = p.1,
            };

            for adj_node in 0..adj.len() {
                let wt = adj[node as usize][adj_node];

                if dis[adj_node] > dis[node as usize] + wt {
                    dis[adj_node] = dis[node as usize] + wt;
                    st.push((dis[adj_node], adj_node as i32));
                }
            }
        }

        return dis;
    }

    fn get_shortest_path_dag(roads: &Vec<Vec<i32>>, dis: &Vec<i128>) -> (Vec<Vec<i32>>, Vec<i32>) {
        let mut adj = vec![Vec::new()as Vec<i32>; dis.len() as usize];
        let mut parent_count = vec![0 as i32; dis.len() as usize];

        for road in roads {
            let u = road[0];
            let v = road[1];
            let wt = road[2];

            if dis[u as usize] + wt as i128 == dis[v as usize] {
                parent_count[v as usize] += 1;
                adj[u as usize].push(v);

            } else if dis[v as usize] + wt as i128 == dis[u as usize] {
                parent_count[u as usize] += 1;
                adj[v as usize].push(u);
            }
        }

        return (adj, parent_count);
    }

    fn topological_sorting(parent_count: &mut Vec<i32>, adj: &Vec<Vec<i32>>) -> Vec<i32> {
        let mut sorted: Vec<i32> = Vec::new();
        let mut q: VecDeque<i32> = VecDeque::new();

        for i in 0..parent_count.len() {
            if parent_count[i] == 0 {
                q.push_back(i as i32);
            }
        }

        while !q.is_empty() {
            let p = q.front();
            let node;
            match p {
                None => break,
                Some(p) => node = *p,
            };

            q.pop_front();
            sorted.push(node);

            for &adj_node in &adj[node as usize] {
                parent_count[adj_node as usize] -= 1;

                if (parent_count[adj_node as usize]) == 0 {
                    q.push_back(adj_node);
                }
            }
        }

        return sorted;
    }

    fn number_of_paths(src: i32, dst: i32, adj: &Vec<Vec<i32>>, sorted: &Vec<i32>) -> i32 {
        let mut dp = vec![0; adj.len() as usize];

        dp[dst as usize] = 1;

        for i in (0..sorted.len()).rev() {
            for j in 0..adj[sorted[i] as usize].len() {
                dp[sorted[i] as usize] = (dp[sorted[i] as usize] + dp[adj[sorted[i] as usize][j] as usize]) % MOD;
            }
        }

        return dp[src as usize];
    }
}
